### PR TITLE
added a script to update llvm-mc test file

### PR DESCRIPTION
**Pull request by broxigarchen**
Added a script to update the test file generated by llvm-mc binary.  The script accepts .s and .txt for asm and dasm.

For mc test I am targetting there is no function name which can be used as a key, thus no clear mapping between input and output. The script assumes the test are always line-by-line and it update the output marker for each test line-by-line.

### CODE REVIEW

**Comment by broxigarchen** [comment]
Hi reviewers, the script is being added when I am working on the amdgpu development, I am no sure how useful this script is, but posting it for review and collect feedbacks.

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm.s.expected
Comment (kosarev) [inline llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm.s.expected:5]: In asm tests we normally put check lines after instructions. Do you propose to change this? (I wouldn't mind either way.)

**Review by kosarev — COMMENTED** [review]
Nice idea.

Does this work for combined asm/disasm tests, e.g., `gfx12_asm_vop1.s`? Does this work for other backend's asm/disasm tests?

**Review by arichardson — COMMENTED** [review]
This looks really useful! I wonder if we could also make it work for the error message checking?

E.g. generate the following check lines
`CHECK-ERR: [[#@LINE]]:<col>: error: ...`

if the RUN: line contains something like `not llvm-mc .... 2>&1 | FileCheck`?

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm.s.expected
Comment (arichardson) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm.s.expected:5]: It probably doesn't matter but I agree for ASM tests having it in the next line (or on the same line) makes more sense to me.

Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm.s.expected:5]: I can put it to the next line. I think putting it on the same line might be a bit too long to read sometime

File: llvm/utils/update_mc_test_check.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_check.py]: Updated. Thanks for the comment!

**Comment by broxigarchen** [comment]
> Nice idea.
> 
> Does this work for combined asm/disasm tests, e.g., `gfx12_asm_vop1.s`? Does this work for other backend's asm/disasm tests?

I tried gfx12_asm_vop1.s, but seems the script does not understand the %extract_encoding% in the cmd line. I think it requires some parsing for these special tokens.

Regarding to the other backend, I did a quick look, and it seems some tests are using ".section" as test seperators so there must be lot of cases that are not covered.

**Comment by broxigarchen** [comment]
> This looks really useful! I wonder if we could also make it work for the error message checking?
> 
> E.g. generate the following check lines `CHECK-ERR: [[#@LINE]]:<col>: error: ...`
> 
> if the RUN: line contains something like `not llvm-mc .... 2>&1 | FileCheck`?

I was looking at it. It's currently not supporting since the `not` is interrupting the binary check. I'll try to see if I can get this to work

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm.s.expected
Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm.s.expected:4]: done

**Comment by broxigarchen** [comment]
> > This looks really useful! I wonder if we could also make it work for the error message checking?
> > E.g. generate the following check lines `CHECK-ERR: [[#@LINE]]:<col>: error: ...`
> > if the RUN: line contains something like `not llvm-mc .... 2>&1 | FileCheck`?
> 
> I was looking at it. It's currently not supporting since the `not` is interrupting the binary check. I'll try to see if I can get this to work

Added

**Review by kosarev — APPROVED** [review]
LVGTM. Please also wait for Alexander's approval.

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected
Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected:5]: I think this can work. Let me try it

Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected]: added

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected
Comment (arichardson) [inline llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected:5]: Can we match the actual column from the error message?

Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected:5]: Hi Alex are you talking about the actual line number here? I made it this way since I found most of the test are using this regex to match line number.

Comment (arichardson) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected:5]: No I meant the column number, i.e. the second component - using `[[#@LINE-1]]` is good since it means no changes if new lines are added.

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:64]:
```suggestion
    if not err:
```

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:56]:
```suggestion
    if not err:
```

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:60]:
```suggestion
    return ERROR_RE.search(err) is not None
```

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:48]:
```suggestion
    line = input_line.strip()
    # Skip empty and comment lines
    if not line or line.startswith(COMMENT[mc_mode]):
        return False
```

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:69]:
```suggestion
    # take the first match
    for line in err.splitlines():
```

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:82]:
```suggestion
    output = ""

    for line in out.splitlines():
```

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:104]:
```suggestion
    output = ""
    for line in output.splitlines():
```

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:116]:
```suggestion
            + ":[[#@LINE-1]]:{{[0-9]+}}: "
```

Comment (arichardson) [reply llvm/utils/update_mc_test_checks.py:116]: Would be good to use the captured column number here instead of {{[0-9]+}} as well

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:21]:
```suggestion
ERROR_RE = re.compile(r"\d+:(\d+): (warning|error): .*")
```
Something like this should allow capturing the error column

**Review by arichardson — COMMENTED** [review]
A few comments inline - looking forward to being able to use this script.

File: llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected
Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected:5]: oh, this might requires more changes. The current approach here is to extract and test line-by-line, so the line number extracted from this python script is not the actual line number when running in llvm-lit

Comment (broxigarchen) [reply llvm/test/tools/UpdateTestChecks/update_mc_test_checks/Inputs/amdgpu_asm_err.s.expected]: Added

**Comment by broxigarchen** [comment]
Quick ping! This PR should be ready to get in

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:73]:
```suggestion
```

This can be omitted (unless out can be None?) since  splitlines returns an empty list in that case and the loop never executes.

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:62]:
```suggestion
```
This can be omitted (unless err can be None?) since  splitlines returns an empty list in that case and the loop never executes.

File: llvm/utils/update_mc_test_checks.py
Comment (arichardson) [inline llvm/utils/update_mc_test_checks.py:56]:
```suggestion
    return err and ERROR_RE.search(err) is not None
```
Just noticed that it can be simplified even more.

**Review by arichardson — APPROVED** [review]
Thanks, this LGTM.

A few minor simplification suggestions if you think this makes it better.

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:62]: Yes err can be None so we cannot omit this

File: llvm/utils/update_mc_test_checks.py
Comment (broxigarchen) [reply llvm/utils/update_mc_test_checks.py:73]: same as above

**Pull request merged**
